x/connectivity/connectivity.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connectivity
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
+	"github.com/miekg/dns"
+)
+
+// TestError captures the observed error of the connectivity test.
+type TestError struct {
+	// Which operation in the test that failed: "dial", "write" or "read"
+	Op string
+	// The POSIX error, when available
+	PosixError string
+	// The error observed for the action
+	Err error
+}
+
+var _ error = (*TestError)(nil)
+
+func (err *TestError) Error() string {
+	return fmt.Sprintf("%v: %v", err.Op, err.Err)
+}
+
+func (err *TestError) Unwrap() error {
+	return err.Err
+}
+
+// TestStreamDialerConnectivity uses the given [transport.StreamDialer] to resolve the test domain using the given DNS resolver.
+// The context can be used to set a timeout or deadline, or to pass values to the dialer.
+func TestStreamDialerConnectivity(ctx context.Context, dialer transport.StreamDialer, resolverAddress string, testDomain string) (time.Duration, error) {
+	dial := func(ctx context.Context, address string) (net.Conn, error) { return dialer.Dial(ctx, resolverAddress) }
```

Comment:
Shall we use `address` instead of `resolverAddress` here?
```suggestion
	dial := func(ctx context.Context, address string) (net.Conn, error) { return dialer.Dial(ctx, address) }
```

Reply:
Also, why not passing `dialer.Dial` directly to `testDialer`?